File: hydra-clojure/src/main/clojure/hydra/lib/pairs.clj
(ns hydra.lib.pairs)

;; Pairs are 2-element lists: (list a b)

;; bimap :: (a -> c) -> (b -> d) -> Pair a b -> Pair c d
(def hydra_lib_pairs_bimap
  (fn [f] (fn [g] (fn [p] (list (f (first p)) (g (second p)))))))

;; first :: Pair a b -> a
(def hydra_lib_pairs_first
  (fn [p] (first p)))

;; second :: Pair a b -> b
(def hydra_lib_pairs_second
  (fn [p] (second p)))

File: overlay/java/hydra-kernel/src/main/java/hydra/dsl/prims/Pairs.java
package hydra.dsl.prims;

import hydra.core.Term;
import hydra.lib.pairs.Bimap;
import hydra.lib.pairs.First;
import hydra.lib.pairs.Second;

/**
 * DSL interface providing pair primitive operations.
 */
public interface Pairs {
    /**
     * Returns a term representing the bimap primitive operation for pairs.
     * Applies two functions to the first and second elements of a pair respectively.
     *
     * @return a term for bimap over pairs
     */
    static Term bimap() {
        return new Bimap().term();
    }

    /**
     * Returns a term representing the first primitive operation for pairs.
     * Extracts the first element of a pair.
     *
     * @return a term for extracting the first element
     */
    static Term first() {
        return new First().term();
    }

    /**
     * Returns a term representing the second primitive operation for pairs.
     * Extracts the second element of a pair.
     *
     * @return a term for extracting the second element
     */
    static Term second() {
        return new Second().term();
    }
}
